bigtable-hbase-1.x-parent/bigtable-hbase-1.x-mapreduce/src/test/java/com/google/cloud/bigtable/mapreduce/validation/EndToEndIT.java

```diff
@@ -0,0 +1,429 @@
+    }
+    return splitRows;
+  }
+
+  public static String getZkConnArgFromConfig(Configuration conf) {
+    return new StringBuilder()
+        .append(conf.get(HConstants.ZOOKEEPER_QUORUM))
+        .append(":")
+        .append(conf.get(HConstants.ZOOKEEPER_CLIENT_PORT))
+        .append(":")
+        .append(conf.get(HConstants.ZOOKEEPER_ZNODE_PARENT))
+        .toString();
+  }
+
+  /** validate data migrated from Bigtable to HBase */
+  @Test
+  public void testSyncTableBigtabletoHBaseValidation() throws Exception {
+    long testRunTs = System.currentTimeMillis();
+
+    // source and target
+    TableName BT_TEST_SOURCE_TABLE = TableName.valueOf("cbt-testsourcetable-" + testRunTs);
+    TableName HB_TEST_TARGET_TABLE = TableName.valueOf("hbase-testtargettable-" + testRunTs);
+    Path testHashOutDir = HB_TEST_UTIL.getDataTestDirOnTestFS("hash-out-cbt-" + testRunTs);
+
+    // generate data
+    writeTestData(BT_TEST_SOURCE_TABLE, false, HB_TEST_TARGET_TABLE, true);
+
+    // hash table
+    hashSourceBigtable(BT_TEST_SOURCE_TABLE, testHashOutDir);
+
+    // sync table validation
+    Counters syncCounters =
+        bigtableSyncTableJobSourceBigtableTargetHBase(
+            BT_TEST_SOURCE_TABLE, HB_TEST_TARGET_TABLE, testHashOutDir);
+
+    assertVerifySyncMatches(syncCounters);
+  }
+
+  /** validate data migrated from HBase to Bigtable */
+  @Test
+  public void testSyncTableHBaseToBigtableValidation() throws Exception {
+    long testRunTs = System.currentTimeMillis();
+
+    // source and target
+    TableName HB_TEST_SOURCE_TABLE = TableName.valueOf("hbase-testsourcetable-" + testRunTs);
+    TableName BT_TEST_TARGET_TABLE = TableName.valueOf("cbt-testtargettable-" + testRunTs);
+    Path testHashOutDir = HB_TEST_UTIL.getDataTestDirOnTestFS("hash-out-hbase-" + +testRunTs);
+
+    // generate data
+    writeTestData(HB_TEST_SOURCE_TABLE, true, BT_TEST_TARGET_TABLE, false);
+
+    // hash table
+    hashSourceHBaseTable(HB_TEST_SOURCE_TABLE, testHashOutDir);
+
+    // sync table validation
+    Counters syncCounters =
+        bigtableSyncTableJobSourceHBaseTargetBigtable(
+            HB_TEST_SOURCE_TABLE, BT_TEST_TARGET_TABLE, testHashOutDir);
+
+    assertVerifySyncMatches(syncCounters);
+  }
+
+  /** assert match/mismatch results from counters and output records */
+  private void assertVerifySyncMatches(Counters syncCounters) {
+    // verify counters
+    assertEquals(60, syncCounters.findCounter(Counter.ROWSWITHDIFFS).getValue());
+    assertEquals(10, syncCounters.findCounter(Counter.SOURCEMISSINGROWS).getValue());
+    assertEquals(10, syncCounters.findCounter(Counter.TARGETMISSINGROWS).getValue());
+    assertEquals(50, syncCounters.findCounter(Counter.SOURCEMISSINGCELLS).getValue());
+    assertEquals(50, syncCounters.findCounter(Counter.TARGETMISSINGCELLS).getValue());
+    assertEquals(20, syncCounters.findCounter(Counter.DIFFERENTCELLVALUES).getValue());
```

Comment:
Where are we checking on the matches?

Reply:
added two additional junit tests to verify matches `testSyncTableBigtabletoHBaseValidationMatches` and `testSyncTableHBaseToBigtableValidationMatches`